analyzers/ImpactParameter.py

```diff
@@ -1,48 +1,214 @@
+                            ipsig_ptcs.append([ptc.path.IP_signif, ptc])
+                            
+                            if self.tag_IP_b_LL:
+                                #import pdb; pdb.set_trace()
+                                ibin = self.ratio_IP.FindBin(ptc.path.IP)
+                                lhratio = self.ratio_IP.GetBinContent(ibin)
+                                if not lhratio == 0:
+                                    ptc.path.IP_b_LL = math.log(lhratio)
+                                    IP_b_LL += ptc.path.IP_b_LL
+                                if lhratio == 0:
+                                    ptc.path.IP_b_LL = 0
+                            
+                            if self.tag_IPs_b_LL:
+                                #import pdb; pdb.set_trace()
+                                ibin = self.ratio_IPs.FindBin(ptc.path.IP_signif)
+                                lhratio = self.ratio_IPs.GetBinContent(ibin)
+                                if not lhratio == 0:
+                                    ptc.path.IPs_b_LL = math.log(lhratio)
+                                    IPs_b_LL += ptc.path.IPs_b_LL
+                                if lhratio == 0:
+                                    ptc.path.IPs_b_LL = 0
+                            
+            ipsig_ptcs.sort(reverse=True)
+            
+            if len(ipsig_ptcs) < 2 :
+                TCHE = -99
+                TCHP = -99
+                TCHE_IP = -99
+                TCHP_IP = -99
+                TCHE_x = -99
+                TCHE_y = -99
+                TCHE_z = -99
+                TCHE_pt = -99
+                TCHE_dr = -99
+                
+            if len(ipsig_ptcs) > 1 :
+                TCHE = ipsig_ptcs[1][0]
+                ptc = ipsig_ptcs[1][1]
+                TCHE_IP = ptc.path.IP
+                TCHE_x, TCHE_y, TCHE_z = ptc.path.coord_at_time(0)
+                TCHE_pt = ptc.path.p4.Perp()
+                TCHE_dr = deltaR(jet.eta(), jet.phi(), ptc.eta(), ptc.phi())
+                TCHP = -99
+                TCHP_IP = -99
+                
+            if len(ipsig_ptcs) > 2 :
+                TCHP = ipsig_ptcs[2][0]
+                ptc = ipsig_ptcs[2][1]
+                TCHP_IP = ptc.path.IP
+                
+            jet.tags['IP_b_LL'] = IP_b_LL  if self.tag_IP_b_LL  else None
+            jet.tags['IPs_b_LL']= IPs_b_LL if self.tag_IPs_b_LL else None
+            jet.tags['TCHE'] = TCHE
+            jet.tags['TCHP'] = TCHP
+            jet.tags['TCHE_IP'] = TCHE_IP
+            jet.tags['TCHP_IP'] = TCHP_IP
+            jet.tags['TCHE_x'] = TCHE_x
+            jet.tags['TCHE_y'] = TCHE_y
+            jet.tags['TCHE_z'] = TCHE_z
+            jet.tags['TCHE_xy'] = (TCHE_x**2+TCHE_y**2)**0.5
+            jet.tags['TCHE_pt'] = TCHE_pt
+            jet.tags['TCHE_dr'] = TCHE_dr
+            
+            if hasattr(event, 'K0s') == True :
```

Comment:
Not sure I understand what I am supposed to do. We may discuss more about it.
I have taken almost all your other comments (I check if it is OK), maybe I make a new pull before we continue on this point ?